src/renderer/components/+namespaces/namespace-select-filter.tsx

```diff
@@ -65,69 +85,90 @@ export class NamespaceSelectFilter extends React.Component<SelectProps> {
-    NamespaceSelectFilter.isMultiSelection.set(val);
+    NonInjectedNamespaceSelectFilter.isMultiSelection.set(val);
   }
 
   @computed get isMenuOpen() {
-    return NamespaceSelectFilter.isMenuOpen.get();
+    return NonInjectedNamespaceSelectFilter.isMenuOpen.get();
   }
 
   set isMenuOpen(val: boolean) {
-    NamespaceSelectFilter.isMenuOpen.set(val);
+    NonInjectedNamespaceSelectFilter.isMenuOpen.set(val);
+  }
+
+  get maxItems() {
+    return this.props.maxItems;
   }
 
   componentDidMount() {
     disposeOnUnmount(this, [
       reaction(() => this.isMenuOpen, newVal => {
         if (newVal) { // rising edge of selection
-          this.selected.replace(namespaceStore.selectedNames);
+          if (this.namespaceStore.areAllSelectedImplicitly) {
+            this.selected.replace([""]);
+          } else {
+            this.selected.replace(this.namespaceStore.selectedNames);
+          }
           this.didToggle = false;
         }
       }),
     ]);
   }
 
   formatOptionLabel({ value: namespace, label }: SelectOption) {
-    if (namespace) {
-      const isSelected = namespaceStore.hasContext(namespace);
-
-      return (
-        <div className="flex gaps align-center">
-          <Icon small material="layers" />
-          <span>{namespace}</span>
-          {isSelected && <Icon small material="check" className="box right" />}
-        </div>
-      );
-    }
+    const isSelected = namespace
+      ? !this.namespaceStore.areAllSelectedImplicitly && this.namespaceStore.hasContext(namespace)
+      : this.namespaceStore.areAllSelectedImplicitly;
 
-    return label;
+    return (
+      <div className="flex gaps align-center">
+        <Icon small material={ namespace ? "layers" : "panorama_wide_angle" } />
+        <span>{label}</span>
+        {isSelected && <Icon small material="check" className="box right" />}
+      </div>
+    );
   }
 
   @action
   onChange = ([{ value: namespace }]: SelectOption[]) => {
     if (namespace) {
       if (this.isMultiSelection) {
-        this.didToggle = true;
-        namespaceStore.toggleSingle(namespace);
+        if (this.inputValue === "" && (this.namespaceStore.items.length < this.maxItems)) {
+          this.didToggle = true;
+          this.namespaceStore.toggleSingle(namespace);
+        } else {
+          if (this.namespaceStore.areAllSelectedImplicitly && namespace === "") {
+            this.namespaceStore.selectSingle(namespace);
+          } else {
+            this.namespaceStore.toggleSingle(namespace);
+          }
+        }
+
       } else {
-        namespaceStore.selectSingle(namespace);
+        this.namespaceStore.selectSingle(namespace);
       }
     } else {
-      namespaceStore.selectAll();
+      this.namespaceStore.selectAll();
     }
   };
 
+  onInputChange = (value: string, meta: InputActionMeta) => {
+    if (meta.action === "input-change") this.inputValue = value;
+    if (meta.action === "menu-close") this.inputValue = "";
```

Comment:
Would be better if we either used a `switch` or use `else if`.

Either way, I think from a readability standpoint using brackets here would be better.